### ISSUE TITLE
We need your help! Typesafe GL handles.

### ISSUE THREAD

**Issue by frederikja163** [opentk/opentk#1297]
## What are typesafe handles?
Typesafe handles means we wanna make the handles typesafe. So instead of using ints for handles, we use different structs for different types of handles. This is how it could look.

Currently (pseudo-code):
```cs
uint buffer = GL.GenBuffer();
GL.BindBuffer(buffer);
// ...
GL.BindBuffer(0);

uint vao = GL.GenVertexArray();
GL.BindVertexArray(vao);

GL.BindBuffer(buffer);
// ...
GL.BindBuffer(0);

GL.BindVertexArray(0);
```

With typesafe handles (pseudo-code).
```cs
Buffer buffer = GL.GenBuffer();
GL.BindBuffer(buffer);
// ...
GL.BindBuffer(0);

VertexArray vao = GL.GenVertexArray();
GL.BindVertexArray(vao);

GL.BindBuffer(buffer);
// ...
GL.BindBuffer(0);

GL.BindVertexArray(0);
```

This means that `vao = buffer;` would succeed in the first example, but it wouldn't in the second example because buffer and vao are of two different types. They are typesafe. It also means we can add some sort of safety so `vao += 1` wont happen.

We need help with the names of these handles though. There is a handle type called Program and one called Shader, these are very likely to be used elsewhere and in order to avoid name conflicts we either need to change the names entirely. Or append/prepend them with some other string. Read the different suggestions for this under *Type names*.

Note all the things in here are only examples, and you can come up with your own example and share them with us, either in the comments on this issue. Or on our discord server https://discord.gg/6HqD48s. Where we will shortly start a discussion over voice text, and over the text chats on what choices to pick. Finally we will also have a vote on the discord server picking between thee different options. Depending on how the vote is held, we might also post it here in the comments.

This is a list of all the typesafe handles we can generate automatically at the moment. Let us know if there are any we have missed and we might be able to get those in aswell.
* Program
* ProgramPipeline
* Texture
* Buffer
* Shader
* Query
* Framebuffer
* Renderbuffer
* Sampler
* TransformFeedback
* VertexArray
* Sync
* DisplayList

We will in this document use GenVertexArray, and the VertexArray types to document and discuss the ideas.

### Current Proposals
Ideas and proposals for name postfixes and suffixes. Please help us think of more!
* GLVertexArray
* GL.VertexArray
* HandleVertexArray
* VertexArrayHandle

## Notes
* This will only affect opentk version 5.x and forward, as the new bindings generator is required for these typesafe handles. But even if you aren't an opentk 5 user and wont be, we still value your feedback a lot.

**Comment by amulware** [comment]
I like this!

Note that instead of allowing `GL.BindBuffer(0)` (either through an overload or implicit int cast, both of which allow messing with the integers), `Gl.BindBuffer(default)` reads really well, and remains completely type safe.

With regards to naming I don't have strong opinions, though some sort of suffix or prefix would be nice, at least for someone like me who already [wraps the handle in their own type](https://github.com/beardgame/graphics/blob/master/Bearded.Graphics/Core/Buffer.cs).

**Comment by frederikja163** [comment]
> I like this!
> 
> Note that instead of allowing `GL.BindBuffer(0)` (either through an overload or implicit int cast, both of which allow messing with the integers), `Gl.BindBuffer(default)` reads really well, and remains completely type safe.
> 
> With regards to naming I don't have strong opinions, though some sort of suffix or prefix would be nice, at least for someone like me who already [wraps the handle in their own type](https://github.com/beardgame/graphics/blob/master/Bearded.Graphics/Core/Buffer.cs).

Lots of people create abstractions like that. This is excactly why we want some sort of prefix. Good idea about the default too. I personally didn't think of doing default. But that will definitely be possible. We were coming to a conclusion of doing something similar to `GL.BindBuffer(Buffer.Zero);`. If you want to come and participate in the discussion. We will hold a meeting tomorrow at 18:00 UTC. (Saturday the 5th of june). At our discord server.

**Comment by deccer** [comment]
Frankly, I would keep `0` as is. Its well known, and well documented in all sorts of GL related topics/discussions/forums/documentations.

We are not trying to abstract anything with OpenTK, But all of us abstracting OpenTK/OpenGL are free to do whatever we want with implementations like `Buffer.cs` etc

**Comment by dfkeenan** [comment]
Are these handles going to derive from `System.Runtime.InteropServices.SafeHandle`?

I have seen multiple name convention in Microsofts' source code. 
* `Safe<Something>Handle`
* `<Something>SafeHandle`

Not sure if the word `Safe` is necessary. Something like `TextureHandle` etc. might be good enough.

**Comment by NogginBops** [comment]
@dfkeenan What `SafeHandle` provides is out of scope for this PR. A proper implementation of `SafeHandle` would require a quite complicated implementation (opengl doesn't like calls from other threads and finalizers are called on their own thread), which we are not aiming for. The implementation of these handles will be simple structs that are castable to `int` or `uint` and doesn't do anything more than just a typesafe wrapper around that int.

`SafeHandle` would also require quite an opinionated implementation where OpenTK would have to call opengl functions that are not directly visible or controllable from user code. OpenTK doesn't aim to provide such abstractions.

So we leave users to implement the `SafeHandle` approach themselves. Which allows people who don't want to use `SafeHandle` to not use it.

For example I would not use a `SafeHandle` implementation because I would like to avoid the pointer chasing that it would nessecary create (as it is a reference type).

**Comment by dfkeenan** [comment]
@NogginBops , yeah that is cool. I had no idea on how this was going to be implemented. Agreed `SafeHandle` is not necessary. In that case I would go with something like `<Something>Handle`.

**Comment by frederikja163** [comment]
We're currently having a meeting on the discord about this issue. Come join us https://discord.gg/6HqD48s

**Comment by frederikja163** [comment]
![image](https://user-images.githubusercontent.com/20968096/122276869-876f8f00-cee5-11eb-9b80-62850d79434e.png)
This is the result of the meeting. It has been implemented in #1303 but we are currently trying to solve some appveyor issues, after those have been resolved we are good to go for a pre-release of otk 5.0 with this feature!
